File: solutions/ocaml/089-maximum-product-subarray.ml
(*
  Problem 89: Maximum Product Subarray (LeetCode 152)
  Difficulty: Med
  Language: OCaml
*)
let max_product nums =
  let hd = List.hd nums in
  let (best, _, _) = List.fold_left (fun (best, mx, mn) n ->
    let hi = max n (max (mx * n) (mn * n)) in
    let lo = min n (min (mx * n) (mn * n)) in
    (max best hi, hi, lo)
  ) (hd, hd, hd) (List.tl nums) in
  best

File: solutions/scala/089-maximum-product-subarray.scala
/**
 * Problem 89: Maximum Product Subarray (LeetCode 152)
 * Difficulty: Med
 * Language: Scala
 */
def maxProduct(nums: List[Int]): Int = {
  val (best, _, _) = nums.tail.foldLeft((nums.head, nums.head, nums.head)) {
    case ((best, mx, mn), n) =>
      val hi = Seq(n, mx * n, mn * n).max
      val lo = Seq(n, mx * n, mn * n).min
      (best max hi, hi, lo)
  }
  best
}
